egs2/librispeech/discrete_asr1/local/data.sh

```diff
@@ -0,0 +1,452 @@
+
+        nutt=$(<"${feats_dir}/${dset}/"feats.scp wc -l)
+        _nj=$((nj<nutt?nj:nutt))
+
+        key_file="${feats_dir}/${dset}"/feats.scp
+        split_scps=""
+        for n in $(seq ${_nj}); do
+            split_scps+=" ${_logdir}/inference_kmeans.${n}.scp"
+        done
+        # shellcheck disable=SC2086
+        utils/split_scp.pl "${key_file}" ${split_scps}
+
+        ${train_cmd} JOB=1:${_nj} ${_logdir}/inference_pseudo_labels_km${nclusters}.JOB.log \
+            ${python} local/dump_km_label.py \
+                --km_path ${km_dir}/km_${nclusters}.mdl \
+                --in_filetype "mat" \
+                --out_filetype "mat" \
+                "scp:${_logdir}/inference_kmeans.JOB.scp" \
+                "ark,t:${_logdir}/pseudo_labels_km${nclusters}.JOB.txt" || exit 1;
+
+        # concatenate scp files
+        for n in $(seq ${_nj}); do
+            cat ${_logdir}/pseudo_labels_km${nclusters}.${n}.txt || exit 1;
+        done | sed 's/ \[ \| \]//g' > "${_logdir}"/../pseudo_labels_km${nclusters}.txt || exit 1;
+    done
+fi
+
+alignment_phoneme_dir="./data/mfa_phoneme_alignment"
+if [ ${stage} -le 8 ] && [ ${stop_stage} -ge 8 ] && ! [[ " ${skip_stages} " =~ [[:space:]]8[[:space:]] ]]; then
+
+    if [ ! -f ${alignment_phoneme_dir}/dev-clean.tsv ] && [ ! -f ${alignment_phoneme_dir}/dev-other.tsv ]; then
+        log "Stage 8: Downloading MFA from https://zenodo.org/record/2619474#.Y2F3ZewVDu0"
+        mkdir -p ${alignment_phoneme_dir}
+
+        wget \
+            -O ${alignment_phoneme_dir}/librispeech_alignments.zip \
+            https://zenodo.org/record/2619474/files/librispeech_alignments.zip?download=1
+
+        unzip "${alignment_phoneme_dir}/librispeech_alignments.zip" -d "${alignment_phoneme_dir}"
+        python local/dump_librispeech_alignment_from_textgrid.py \
+            --alignment_root "${alignment_phoneme_dir}" \
+            --dataset "dev-other" "dev-clean"
+    else
+        log "Stage 8: Librispeech phoneme alignments exists. Skipping ..."
+    fi
+fi
+
+if [ ${stage} -le 9 ] && [ ${stop_stage} -ge 9 ] && ! [[ " ${skip_stages} " =~ [[:space:]]9[[:space:]] ]]; then
+    log "Stage 9: measure qualities of labels: purity"
+
+    # (simpleoier): For LibriSpeech task only
+    ${python} local/measure_teacher_quality.py \
+        --lab_dir "${feats_dir}" \
+        --lab_name "pseudo_labels_km${nclusters}.txt" \
+        --lab_sets "dev_clean" "dev_other" \
+        --phn_dir "${alignment_phoneme_dir}" \
+        --phn_sets "dev-clean" "dev-other" \
+        --pad_len 0 \
+        --upsample 2 \
+        --ref_lab_dir "" \
+        --ref_lab_name "" | tee ${km_dir}/phoneme_pseudo_label_quality.txt
+
+fi
+
+if [ ${stage} -le 10 ] && [ ${stop_stage} -ge 10 ] && ! [[ " ${skip_stages} " =~ [[:space:]]10[[:space:]] ]]; then
+    log "Stage 10: Prepare token_list and convert number indices to CJK tokens"
+
+    # Get uniq chars
+    if [ ! -f "${km_dir}/../"distinct_cjk_token_lists ]; then
```

Comment:
Just curios, how you would handle the units with more than 5000 then?

Reply:
Good catch. I should change the 4096 to nclusters, and adding a check if it is less than the total number of CJK tokens. I believe we can use up to 10K tokens.